lib/stitches/auto_fill.py

```diff
@@ -128,248 +131,241 @@ def build_graph(shape, segments, angle, row_spacing, max_stitch_length):
+    """Build a graph for travel stitches.
 
-        if path is None:
-            # This is the very first time through the loop, so initialize.
-            path = []
-            ending_node = starting_node
-        else:
-            ending_node = path[-1][-1]
+    This graph will be used to find a stitch path between two spots on the
+    outline of the shape.
 
-        # get a list of neighbors paired with the key of the edge I can follow to get there
-        neighbors = [
-            (node, key)
-            for node, adj in graph.adj[ending_node].iteritems()
-            for key in adj
-        ]
+    If underpath is False, we'll just be traveling
+    around the outline of the shape, so the graph will only contain outline
+    edges.
 
-        # heuristic: try grating segments first
-        neighbors.sort(key=lambda dest_key: dest_key[1] == "segment", reverse=True)
+    If underpath is True, we'll also allow travel inside the shape.  We'll
+    fill the shape with a cross-hatched grid of lines.  We'll construct a
+    graph from them and use a shortest path algorithm to construct travel
+    stitch paths in travel().
 
-        for next_node, key in neighbors:
-            # skip if I've already followed this edge
-            edge = PathEdge((ending_node, next_node), key)
-            if edge in visited_edges:
-                continue
+    When underpathing, we "encourage" the travel() function to travel inside
+    the shape rather than on the boundary.  We do this by weighting the
+    boundary edges extra so that they're more "expensive" in the shortest path
+    calculation.  We also weight the interior edges extra proportional to
+    how close they are to the boundary.
+    """
 
-            new_path = path + [edge]
+    graph = networkx.MultiGraph()
 
-            if next_node == starting_node:
-                # ignore trivial loops (down and back a doubled edge)
-                if len(new_path) > 3:
-                    return new_path
+    # Add all the nodes from the main graph.  This will be all of the endpoints
+    # of the rows of stitches.  Every node will be on the outline of the shape.
+    # They'll all already have their `outline` and `projection` tags set.
+    graph.add_nodes_from(fill_stitch_graph.nodes(data=True))
 
-            new_visited_edges = visited_edges.copy()
-            new_visited_edges.add(edge)
+    if underpath:
+        boundary_points, travel_edges = build_travel_edges(shape, fill_stitch_angle)
 
-            to_search.appendleft((new_path, new_visited_edges))
+        # This will ensure that a path traveling inside the shape can reach its
+        # target on the outline, which will be one of the points added above.
+        tag_nodes_with_outline_and_projection(graph, shape, boundary_points)
 
+    add_edges_between_outline_nodes(graph)
 
-def find_loop(graph, starting_nodes):
-    """find a loop in the graph that is connected to the existing path
+    if underpath:
```

Comment:
I would return early in this function so that the rest of the code isn't indented, i.e., 

```
if not underpath:
     return
```

Reply:
Good idea.  This function got longer than it really should be.  I think I'll try to split it out a bit more and return early as you suggest.